src/kv/ZetaScaleStore.cc

```diff
@@ -0,0 +1,1030 @@
+      if (inclusive)
+	append_chunk_number(skey, skey.length(), MAX_CHUNK_COUNT);
+      meta.key_start = (char *)skey.c_str();
+      meta.keylen_start = skey.length();
+    }
+  } else if (!direction) {
+    /* Handle descending unbounded query, used for seek_to_last() */
+    meta.flags = (ZS_range_enums_t)ZS_RANGE_END_GE;
+    meta.key_end = (char *)smallest;
+    meta.keylen_end = 0;
+  }
+
+  if (snap_seqno != 0) {
+    meta.flags = (ZS_range_enums_t)(meta.flags | ZS_RANGE_SEQNO_LE);
+    meta.end_seq = snap_seqno;
+  }
+
+  status =
+      ZSGetRange(_thd_state(), cguid, ZS_RANGE_PRIMARY_INDEX, &cursor, &meta);
+
+  dtrace << " " << cguid << " " << inclusive << " " << direction << " ["
+	 << decode_key(key ? key : smallest) << "] " << snap_seqno
+	 << " length: " << length << dendl;
+
+  if (status != ZS_SUCCESS) {
+    derr << "ZSGetRange failed: " << ZSStrError(status) << dendl;
+    assert(!cursor);
+  } else if (_next())
+    return -1;
+
+  return (lc_it != NULL || status == ZS_SUCCESS) ? 0 : -1;
+}
+
+int ZSStore::ZSWholeSpaceIteratorImpl::_next()
+{
+  ZS_status_t status;
+  int n_out;
+
+  cur_key.clear();
+
+  if (lc_it != NULL) {
+    char *ekey = NULL, *data = NULL;
+    uint32_t keylen;
+    uint64_t datalen;
+
+    status = ZSNextEnumeratedObject(_thd_state(), lc_it, &ekey, &keylen, &data,
+				    &datalen);
+    if (status == ZS_SUCCESS) {
+      char *p = strchr(ekey + 2, '_') + 1;
+      cur_key.append(p, keylen - (p - ekey));
+      value_bl.clear();
+      value_bl.append(bufferptr(data, datalen));
+      free(ekey);
+      free(data);
+      dtrace << " lc key=" << decode_key(cur_key) << dendl;
+      return 0;
+    }
+
+    dtrace << "finish logging container enumeration" << dendl;
+
+    status = ZSFinishEnumeration(_thd_state(), lc_it);
+    if (status != ZS_SUCCESS)
+      derr << "ZSFinishEnumeration failed: " << ZSStrError(status) << dendl;
+
+    lc_it = NULL;
+  }
+
+  if (!cursor)
+    return -1;
+
+  status = ZSGetNextRange(_thd_state(), cursor, MAX_CHUNK_COUNT - count, &n_out,
+			  values + count);
+  n_out += count;
+
+  bool _valid = (status == ZS_SUCCESS || status == ZS_QUERY_DONE) && n_out > 0;
+
+  dtrace << _valid << " n_out " << n_out << " count=" << count << dendl;
+
+  if (n_out) {
+    std::string key(values[0].key, values[0].keylen);
+    cur_key.assign(values[0].key, values[0].keylen);
+
+    value_bl.clear();
+    value_bl.append(bufferptr(values[0].data, values[0].datalen));
+
+    free(values[0].data);
+    free(values[0].key);
+
+    dtrace << "first key=" << decode_key(cur_key) << dendl;
+
+    key += '\0';
+    int i = 1;
+    while (i < n_out && values[i].keylen == key.length() + 2 &&
+	   !memcmp(key.c_str(), values[i].key, key.length())) {
+      value_bl.append(bufferptr(values[i].data, values[i].datalen));
+      dtrace << "key=" << decode_key(string(values[i].key, values[i].keylen))
+	     << dendl;
+      free(values[i].data);
+      free(values[i].key);
```

Comment:
There is a buffer constructor that takes over ownership of a malloc'd buffer.. see buffer::claim_malloc().  This would avoid the copy.